wp-includes/class-wp-offline-page-excluder.php

```diff
@@ -0,0 +1,125 @@
+<?php
+/**
+ * WP_Offline_Page_Excluder class.
+ *
+ * @package PWA
+ */
+
+/**
+ * Class handles excluding the default offline page's from the backend and frontend.
+ */
+class WP_Offline_Page_Excluder {
+
+	/**
+	 * Instance of the default offline page Manager.
+	 *
+	 * @var WP_Offline_Page
+	 */
+	protected $manager;
+
+	/**
+	 * WP_Offline_Page_Filter constructor.
+	 *
+	 * @param WP_Offline_Page $manager Instance of the manager.
+	 */
+	public function __construct( $manager ) {
+		$this->manager = $manager;
+	}
+
+	/**
+	 * Initializes the instance.
+	 */
+	public function init() {
+		add_filter( 'wp_dropdown_pages', array( $this, 'exclude_from_page_dropdown' ), 10, 2 );
+		add_action( 'parse_query', array( $this, 'exclude_from_query' ) );
+	}
+
+	/**
+	 * Filters the HTML output of a list of pages as a drop down.
+	 *
+	 * @param string $html HTML output for drop down list of pages.
+	 * @param array  $args The parsed arguments array.
+	 *
+	 * @return string Filtered content for wp_dropdown_pages.
+	 */
+	public function exclude_from_page_dropdown( $html, $args ) {
+		// Bail out if this is the offline page dropdown.
+		if ( WP_Offline_Page::OPTION_NAME === $args['name'] ) {
+			return $html;
+		}
+
+		return preg_replace( '/<option .*? value="' . $this->manager->get_offline_page_id() . '">.*?<\/option>/', '', $html );
+	}
+
+	/**
+	 * Exclude the offline page from the query when doing a search on the frontend or on the backend Menus page.
+	 *
+	 * @param WP_Query $query The WP_Query instance.
+	 */
+	public function exclude_from_query( WP_Query $query ) {
+		if ( $this->is_offline_page_query( $query ) ) {
+			add_action( 'wp_head', 'wp_no_robots' );
```

Comment:
I don't think this is quite right because it could result in `wp_no_robots` being added unintentionally. Consider, for example, a request to a normal post on a site which at `init` or some other early hook does (for some reason):

```php
$offline_query = new WP_Query( array(
    'page_id' => get_option( WP_Offline_Page::OPTION_NAME )
) );
```

This will result in the page getting `wp_no_robots` added since the `parse_query` action runs for every instance of `WP_Query` whether it is the main query or not. 

It would be better to separate the concerns here to add a new method like:

```php
function show_no_robots_on_offline_page() {
    global $wp_query;
    if ( $this->is_offline_page_query( $wp_query ) ) {
        wp_no_robots();
    }
}
```

And then add this method to the `wp_head` action unconditionally in `WP_Offline_Page_Excluder::init()`.

Reply:
It is giving a side effect. I was trying to avoid hooking in again and rerunning the check.  But I do like the separation.  Agree on your points.

Reply:
@westonruter Changes added in [commit af9aedd](https://github.com/xwp/pwa-wp/pull/28/commits/af9aedd1c57cff158d5a4b000ef781ce5b054661).